server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SHARDS.java

```diff
@@ -0,0 +1,234 @@
+      final Map<Address, String> responses = new ConcurrentHashMap<>(members.size());
+      return executor.filterTargets(members)
+            .submitConsumer(SHARDS::readLocalNodeInformation, (address, res, t) -> {
+               if (t != null) {
+                  throw CompletableFutures.asCompletionException(t);
+               }
+
+               responses.put(address, res);
+            }).thenApply(ignore -> responses);
+   }
+
+   private static String readLocalNodeInformation(EmbeddedCacheManager ecm) {
+      CacheManagerInfo manager = ecm.getCacheManagerInfo();
+      String name = manager.getNodeName();
+      Address address = findPhysicalAddress(ecm);
+      int port = 0;
+      if (address instanceof JGroupsAddress && ((JGroupsAddress) address).getJGroupsAddress() instanceof IpAddress) {
+         JGroupsAddress jAddress = (JGroupsAddress) address;
+         port = ((IpAddress) jAddress.getJGroupsAddress()).getPort();
+      }
+      String addressString = address != null ? getOnlyIp(address) : ecm.getCacheManagerInfo().getNodeAddress();
+
+      StringBuilder sb = new StringBuilder();
+
+      // Define the number of properties and values we return.
+      // We do not include the role here. This is added on the caller.
+      sb.append("*14\r\n");
+
+      sb.append("$2\r\n").append("id\r\n");
+      sb.append("$").append(name.length()).append("\r\n").append(name).append("\r\n");
+
+      sb.append("$4\r\n").append("port\r\n");
+      sb.append(":").append(port).append("\r\n");
+
+      sb.append("$2\r\n").append("ip\r\n");
+      sb.append("$").append(addressString.length()).append("\r\n").append(addressString).append("\r\n");
+
+      sb.append("$8\r\n").append("endpoint\r\n");
+      sb.append("$").append(addressString.length()).append("\r\n").append(addressString).append("\r\n");
+
+      sb.append("$18\r\n").append("replication-offset\r\n");
+      sb.append(":0\r\n");
+
+      sb.append("$6\r\n").append("health\r\n");
+      sb.append("$6\r\n").append("online\r\n");
+
+      return sb.toString();
+   }
+
+   private static Address findPhysicalAddress(EmbeddedCacheManager ecm) {
+      Transport transport = ecm.getTransport();
+      if (transport == null) {
+         return null;
+      }
+
+      List<Address> addresses = transport.getPhysicalAddresses();
+      if (addresses.isEmpty()) {
+         // Returning a logical address.
+         return ecm.getAddress();
+      }
+      return addresses.get(0);
+   }
+
+   private static String getOnlyIp(Address address) {
+      if (address instanceof JGroupsAddress && ((JGroupsAddress) address).getJGroupsAddress() instanceof IpAddress) {
+         JGroupsAddress jAddress = (JGroupsAddress) address;
+         return ((IpAddress) jAddress.getJGroupsAddress()).getIpAddress().getHostAddress();
+      }
+      return address.toString();
+   }
+
+   private static CharSequence serialize(String leaderInfo, List<String> replicas, IntSet ranges) {
```

Comment:
I would just pass in the StringBuilder as an argument so we don't have to allocate a StringBuilder for every primary/backup combination

Reply:
Updated.